tests/baselines/reference/constructorParametersInVariableDeclarations.errors.txt

```diff
@@ -1,36 +1,36 @@
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(2,17): error TS2304: Cannot find name 'x'.
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(3,22): error TS2304: Cannot find name 'x'.
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(4,23): error TS2304: Cannot find name 'x'.
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(10,17): error TS2304: Cannot find name 'x'.
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(11,22): error TS2304: Cannot find name 'x'.
-tests/cases/compiler/constructorParametersInVariableDeclarations.ts(12,23): error TS2304: Cannot find name 'x'.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(2,17): error TS2301: Initializer of instance member variable 'a' cannot reference identifier 'x' declared in the constructor.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(3,22): error TS2301: Initializer of instance member variable 'b' cannot reference identifier 'x' declared in the constructor.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(4,23): error TS2301: Initializer of instance member variable 'c' cannot reference identifier 'x' declared in the constructor.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(10,17): error TS2301: Initializer of instance member variable 'a' cannot reference identifier 'x' declared in the constructor.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(11,22): error TS2301: Initializer of instance member variable 'b' cannot reference identifier 'x' declared in the constructor.
+tests/cases/compiler/constructorParametersInVariableDeclarations.ts(12,23): error TS2301: Initializer of instance member variable 'c' cannot reference identifier 'x' declared in the constructor.
 
 
 ==== tests/cases/compiler/constructorParametersInVariableDeclarations.ts (6 errors) ====
     class A {
         private a = x;
                     ~
-!!! error TS2304: Cannot find name 'x'.
+!!! error TS2301: Initializer of instance member variable 'a' cannot reference identifier 'x' declared in the constructor.
         private b = { p: x };
                          ~
-!!! error TS2304: Cannot find name 'x'.
+!!! error TS2301: Initializer of instance member variable 'b' cannot reference identifier 'x' declared in the constructor.
         private c = () => x;
                           ~
-!!! error TS2304: Cannot find name 'x'.
+!!! error TS2301: Initializer of instance member variable 'c' cannot reference identifier 'x' declared in the constructor.

```

Comment:
but not this. Maybe the check for TS2301 should not happen in arrow function assignments?

Reply:
I don't think `x` should be in scope for this arrow function; `x` in this example is just a constructor parameter.

I'm a little confused as to why all of these say that it's "declared in the constructor"; my brain is telling me that it's something that's assigned to `this` in the constructor, when it's just some variable that happens to be in scope for the constructor function, so it's a little weird. I would think that "cannot find name x" is technically the more accurate message, because there's nothing you can do to `x` in the constructor signature (not even `private x: number`) that would make it accessible. (But, there is that comment that says "it's legal in this one situation", which is complicating.)

I guess I'd have to go back to when this diagnostic was added to see what the intent was.

Reply:
The reason for the error is that pre-ES2021 (2022?), TS defaults to emitting class properties as assignments in the constructor (controllable separately from `target` with `useDefinedForClassFields`), so `private a = x` would emit:

```js
class A {
  constructor(x) {
    this.a = x
  }
}
```

Then there are two problems:

1. The thing you wrote *should* not find name 'x' -- so you'd expect the old error.
2. But it *will* at runtime, because the emitted code in the constructor -- so you'd expect the new error.

The new error message is correct when `useDefineForClassFields: false`, which it is for this file, ***and also*** extremely obscure.

Note that you can get the new error without the old one if you add `var x = 1` at the beginning of the test. The new error is possibly even more confusing in this case.

So: the new error is correct and also horrible. I don't know if we should fix it (separately) though: the top hit for the error is a pretty good explanation from 2015. And over time fewer people will target non-compliant emit.

Minutes later: The error still shows up when useDefineForClassFields: true, which it shouldn't. I'll file a bug for that.

Reply:
Filed #49719